File: rust/tcl-registry/src/commands/irules/gtp__payload.rs
// tcl-lsp — a language server and toolchain for Tcl
// Copyright (C) 2026 James Deucker (bitwisecook) <https://github.com/bitwisecook>
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//
// SPDX-License-Identifier: AGPL-3.0-or-later

//! `GTP::payload` iRules command.
use crate::prelude::*;
pub const fn spec() -> CommandSpec {
    CommandSpec {
        name: "GTP::payload",
        dialects: Some(DialectSet::IRULES),
        arity: Arity::at_least(0),
        hover: Some(HoverSnippet {
            summary: "Returns the entire payload for G-PDU message.",
            synopsis: &[
                "GTP::payload",
                "GTP::payload COUNT",
                "GTP::payload OFFSET COUNT",
                "GTP::payload 'replace' ('-message' MESSAGE)? OFFSET COUNT NEW_VALUE",
            ],
            snippet: "Returns the payload, either complete or partial, for G-PDU message.\nThis command returns an empty value, in case of non-G-PDU messages.",
            source: "https://clouddocs.f5.com/api/irules/GTP__payload.html",
            examples: "when CLIENT_ACCEPTED {\n    set payload [UDP::payload]\n    set t2 [GTP::parse $payload]\n    log local0. \"GTP version [GTP::header version -message $t2]\"\n    log local0. \"GTP type [GTP::header type -message $t2]\"\n    log local0. \"GTP teid [GTP::header teid -message $t2]\"\n}",
            return_value: "",
        }),
        forms: &[FormSpec {
            kind: FormKind::Default,
            synopsis: "GTP::payload",
            dialects: None,
        }],
        options: const {
            &[OptionSpec {
                name: "-message",
                value: OptionValue::value(""),
                detail: "Option -message.",
                dialects: None,
                aliases: &[],
                lifecycle: Lifecycle::UNSPECIFIED,
                min_abbrev: None,
            }]
        },
        side_effects: &[SideEffect {
            target: SideEffectTarget::NetworkIo,
            reads: true,
            writes: false,
            connection_side: ConnectionSide::Both,
            dialects: None,
        }],
        taint_source: Some(TaintColour::TAINTED),
        byte_array_payload: Some(BytePayloadSpec {
            message_flag_shift: true,
            ..BytePayloadSpec::DEFAULT
        }),
        data_collection: Some(GTP_PAYLOAD),
        ..CommandSpec::DEFAULT
    }
}

File: core/commands/registry/irules/gtp__payload.py
# Enriched from F5 iRules reference documentation.
"""GTP::payload -- Returns the entire payload for G-PDU message."""

from __future__ import annotations

from ....compiler.side_effects import ConnectionSide, SideEffect, SideEffectTarget
from .._base import CommandDef, make_av
from ..models import (
    CommandSpec,
    FormKind,
    FormSpec,
    HoverSnippet,
    OptionSpec,
    ValidationSpec,
)
from ..namespace_models import EventRequires
from ..signatures import Arity
from ..taint_hints import TaintColour, TaintHint
from ._base import _IRULES_ONLY, register

_SOURCE = "https://clouddocs.f5.com/api/irules/GTP__payload.html"


_av = make_av(_SOURCE)


@register
class GtpPayloadCommand(CommandDef):
    name = "GTP::payload"

    @classmethod
    def spec(cls) -> CommandSpec:
        return CommandSpec(
            name="GTP::payload",
            dialects=_IRULES_ONLY,
            hover=HoverSnippet(
                summary="Returns the entire payload for G-PDU message.",
                synopsis=(
                    "GTP::payload",
                    "GTP::payload COUNT",
                    "GTP::payload OFFSET COUNT",
                    "GTP::payload 'replace' ('-message' MESSAGE)? OFFSET COUNT NEW_VALUE",
                ),
                snippet=(
                    "Returns the payload, either complete or partial, for G-PDU message.\n"
                    "This command returns an empty value, in case of non-G-PDU messages."
                ),
                source=_SOURCE,
                examples=(
                    "when CLIENT_ACCEPTED {\n"
                    "    set payload [UDP::payload]\n"
                    "    set t2 [GTP::parse $payload]\n"
                    '    log local0. "GTP version [GTP::header version -message $t2]"\n'
                    '    log local0. "GTP type [GTP::header type -message $t2]"\n'
                    '    log local0. "GTP teid [GTP::header teid -message $t2]"\n'
                    "}"
                ),
            ),
            forms=(
                FormSpec(
                    kind=FormKind.DEFAULT,
                    synopsis="GTP::payload",
                    options=(
                        OptionSpec(name="-message", detail="Option -message.", takes_value=True),
                    ),
                    arg_values={
                        0: (
                            _av(
                                "replace",
                                "GTP::payload replace",
                                "GTP::payload 'replace' ('-message' MESSAGE)? OFFSET COUNT NEW_VALUE",
                            ),
                        )
                    },
                ),
            ),
            validation=ValidationSpec(
                arity=Arity(),
            ),
            event_requires=EventRequires(),
            side_effect_hints=(
                SideEffect(
                    target=SideEffectTarget.NETWORK_IO,
                    reads=True,
                    connection_side=ConnectionSide.BOTH,
                ),
            ),
        )

    @classmethod
    def taint_hints(cls) -> TaintHint:
        return TaintHint(source={None: TaintColour.TAINTED})
